canonical/cloud-init
# feat: check for preserve_etchostname_status key before writing /etc/hostname

## Proposed Commit Message


```
Some users would like to be able to bring up their systems without automatically creating /etc/hostname if it does not exist already. On distros with systemd, the hostname/hostnamectl command can be used to set the hostname transiently (which does not create /etc/hostname).  To support this without changing previous behavior (that cloud-init will set the hostname statically and create /etc/hostname if it does not already exist), this change proposes a new key preserve_etchostname_state which is true, sets the hostname transiently and if not present or False, sets it statically.

This change does not effect BSD or OpenBSD (or any distros that inherit their _write_hostname), as their hostname management does not use an /etc/hostname file.
```

## Test Steps
Currently, this PR only tests that the proposed changes do not impact any default fuctionality and are fully backward compatible.  If the approach and key name are generally agreed upon, I will add additional unit tests verifying that the behaviour with preserve_etchostname_state set to True is as described above and create documentation for the new key.

## Checklist:
